feat(improve-api-endpoints): Added Datasets and Annotation APIs

jasonfish568 opened the pull request:
- Added AnnotationReplyActionApi, AnnotationListApi, AnnotationUpdateDeleteApi
- Added GET and PATCH endpoints to DatasetApi
- Applied current_user in validate_app_token just like what it is done in validate_dataset_token
- Added available model API

Close #11727

# Summary

Basically copied the functions from console to the service api. Also added clear_partial_member_list() to Dataset Delete endpoint as what has been added in the console api.

I will keep working on this feature until it fulfills my needs. More endpoints will be added.

# Checklist

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [ ] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

jasonfish568 committed:
feat(improve-api-endpoints): Added Datasets and Annotation APIs

- Added AnnotationReplyActionApi, AnnotationListApi, AnnotationUpdateDeleteApi
- Added GET and PATCH endpoints to DatasetApi
- Applied current_user in validate_app_token just like what it is done in validate_dataset_token

Related to: #11727

jasonfish568 committed:
feat(improve-api-endpoints): Added available model API

This API endpoint shares the same API token with Dataset API.
It is essential for patching the knowledge base in terms of embedding model.

Related to: #11727

jasonfish568 committed:
feat(improve-api-endpoints): Improve annotation API

Because app token is assigned to a specific app, the validate_app_token decorator has already validated the app_id.
The returned app_model is the app targeted. Therefore there is no need to pass app_id separately.
Simply use app_model.id instead. So app_id is removed from the api routes.

Related to: #11727

crazywoola commented:
Please fix the errors in CI.

jasonfish568 committed:
style: fixed ruff errors

jasonfish568 committed:
style: Fix ruff error

jasonfish568 commented:
> Please fix the errors in CI.

Errors fixed.

crazywoola commented:
> > Please fix the errors in CI.
> 
> Errors fixed.

Please run `dev/reformat` to resolve the errors in the CI.

jasonfish568 committed:
style: Fix lint errors

jasonfish568 commented:
> > > Please fix the errors in CI.
> > 
> > 
> > Errors fixed.
> 
> Please run `dev/reformat` to resolve the errors in the CI.

Sorry for the trouble. Code has now been reformated.

crazywoola commented:
The code LGTM, however you should add according docs to the API Access as well in CN, EN & JA.

jasonfish568 commented:
> The code part LGTM, however you should add according docs to the API Access as well in CN, EN & JA.

https://github.com/langgenius/dify-docs/pull/424

The relevant docs have been added and amended outdated part.

I have included both Chinese and English versions. However as I do not speak Japanese, I wouldn't be able to verify the translated version therefore I would wait for a Japanese contributor get it translated in a later time.

jasonfish568 commented:
> The code part LGTM, however you should add according docs to the API Access as well in CN, EN & JA.

Could you let me know when this PR will be merged? How does dify team manage PRs? When do PRs ususally get merged?

crazywoola commented:
> > The code part LGTM, however you should add according docs to the API Access as well in CN, EN & JA.
> 
> Could you let me know when this PR will be merged? How does dify team manage PRs? When do PRs ususally get merged?

Hello, for newly added apis, you need to update the docs in mdx files as well.

- There is a conflict file.
- Here are the docs I mentioned.

<img width="481" alt="image" src="https://github.com/user-attachments/assets/d70ae2ec-60c9-4107-9486-fc0ca4c94e32" />

<img width="405" alt="image" src="https://github.com/user-attachments/assets/93addf61-a1be-4dfa-b605-cdc559be043b" />

jasonfish568 committed:
Merge branch 'main' into main

jasonfish568 committed:
docs: Added docs in mdx files for the Dataset APIs implemented

jasonfish568 committed:
Merge branch 'main' of github.com:jasonfish568/dify

jasonfish568 committed:
docs: Added annotation API documentation to Chat App documentation.

It is added to Chat App due to Annotation is under Apps, and there is no individual Annotation API doc page. It seems correct to have Annotation APIs put here.

jasonfish568 commented:
@JohnJyong Happy New Year. Just a follow up on this PR. Please kindly process at your earliest convenience.

jasonfish568 commented:
> > > The code part LGTM, however you should add according docs to the API Access as well in CN, EN & JA.
> > 
> > 
> > Could you let me know when this PR will be merged? How does dify team manage PRs? When do PRs ususally get merged?
> 
> Hello, for newly added apis, you need to update the docs in mdx files as well.
> 
> * There is a conflict file.
> * Here are the docs I mentioned.
> 
> <img alt="image" width="481" src="https://private-user-images.githubusercontent.com/100913391/402494384-d70ae2ec-60c9-4107-9486-fc0ca4c94e32.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4MTgwMTksIm5iZiI6MTczODgxNzcxOSwicGF0aCI6Ii8xMDA5MTMzOTEvNDAyNDk0Mzg0LWQ3MGFlMmVjLTYwYzktNDEwNy05NDg2LWZjMGNhNGM5NGUzMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNlQwNDU1MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00Mjg4Mjk4ZTIyNTFjMWI2MTM2YTA1NjQyYzgxZTJhMDUzMTRlMmE3YWRiMzc2YzZjYzc1MzNlNzg4NGFjZTcxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.KWpyOe8lbwgbS11QFjl6e-ICXDPGpckmwhfN6spxmsA"> <img alt="image" width="405" src="https://private-user-images.githubusercontent.com/100913391/402494702-93addf61-a1be-4dfa-b605-cdc559be043b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4MTgwMTksIm5iZiI6MTczODgxNzcxOSwicGF0aCI6Ii8xMDA5MTMzOTEvNDAyNDk0NzAyLTkzYWRkZjYxLWExYmUtNGRmYS1iNjA1LWNkYzU1OWJlMDQzYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNlQwNDU1MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZDZhN2RhNmJhNDMzMDRhMjY1MmQzMzg4YTdlMDhmZmE1Y2U4NGNkZGUyMTE5ZTlhNzE5ZDFjODYyZmNiNTNlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.YRSh7rl3Ftllit_NUtFfdVKB2FTkby9NJRT_Jip8KGQ">

Hi,

I was wondering if you could speed up the merge of this PR? I'm desparately willing to use these APIs for my project.

Thanks

jasonfish568 committed:
Merge pull request #1 from langgenius/main

Update to lastest version

jasonfish568 committed:
Merge pull request #2 from langgenius/main

Update

jasonfishblu commented:
@crazywoola Could you please clarify that version 1.0 has got these APIs implemented? The Dify Doc PR got denied due to not planned.

If this is never gonna be a feature in the official package, then I won't be bothered to contribute. My pay rate is too high to be wasted. Very much disappointed that when other contributors spent quite some time but received no response.

crazywoola commented:
> @crazywoola Could you please clarify that version 1.0 has got these APIs implemented? The Dify Doc PR got denied due to not planned.
> 
> If this is never gonna be a feature in the official package, then I won't be bothered to contribute. My pay rate is too high to be wasted. Very much disappointed that when other contributors spent quite some time but received no response.

Sorry for the late response, this is a great feature that we would like to have.

I know that the explanation is rather insipid. The fact is, we are having a major infra update this week.

- See https://github.com/langgenius/dify/pull/16420

So most of the PRs can not get merged during this time.

Can you resolve the conflicts, I will merged it asap. :)

jasonfishblu commented:
Thanks for the reply, I will arrange a time to get the conflicts sorted.

crazywoola commented:
> Thanks for the reply, I will arrange a time to get the conflicts sorted.

Cool Thanks, sorry for the inconvience.

jasonfish568 committed:
Merge branch 'main' into main

jasonfish568 commented:
> > Thanks for the reply, I will arrange a time to get the conflicts sorted.
> 
> Cool Thanks, sorry for the inconvience.

Hi Crazywoola,

I have managed to get the conflicts resolved. New doc added by other contributors did get honoured so no info missed.

Please help get this PR merged to prevent further conflicts at your earliest convenience.

Very much appreciated!

Jason

jasonfish568 commented on api/controllers/service_api/app/annotation.py:
```diff
@@ -0,0 +1,107 @@
+        app_annotation_job_key = "{}_app_annotation_job_{}".format(action, str(job_id))
+        cache_result = redis_client.get(app_annotation_job_key)
+        if cache_result is None:
+            raise ValueError("The job is not exist.")
+
```
Um....I don't think I wrote this but will create another PR to fix it now.
